File: app/src/main/java/com/absinthe/anywhere_/ui/shortcuts/CreateShortcutDialogFragment.java
package com.absinthe.anywhere_.ui.shortcuts;

import android.app.Dialog;
import android.content.Context;
import android.content.Intent;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;

import com.absinthe.anywhere_.AnywhereApplication;
import com.absinthe.anywhere_.R;
import com.absinthe.anywhere_.model.AnywhereEntity;
import com.absinthe.anywhere_.model.Const;
import com.absinthe.anywhere_.utils.ShortcutsUtils;
import com.absinthe.anywhere_.utils.UiUtils;
import com.absinthe.anywhere_.view.AnywhereDialogBuilder;
import com.absinthe.anywhere_.view.AnywhereDialogFragment;
import com.absinthe.anywhere_.viewbuilder.entity.CreateShortcutDialogBuilder;
import com.bumptech.glide.Glide;
import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions;

import static android.app.Activity.RESULT_OK;

public class CreateShortcutDialogFragment extends AnywhereDialogFragment {
    private Context mContext;
    private CreateShortcutDialogBuilder mBuilder;

    private AnywhereEntity mEntity;
    private Drawable mIcon;
    private String mName;

    public CreateShortcutDialogFragment(AnywhereEntity ae) {
        mEntity = ae;
        mName = ae.getAppName();
        mIcon = UiUtils.getAppIconByPackageName(AnywhereApplication.sContext, ae.getParam1());
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        mContext = getActivity();
    }

    @RequiresApi(api = Build.VERSION_CODES.O)
    @NonNull
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        AnywhereDialogBuilder builder = new AnywhereDialogBuilder(mContext);
        mBuilder = new CreateShortcutDialogBuilder(mContext);
        initView();

        return builder.setView(mBuilder.getRoot())
                .setTitle(R.string.dialog_set_icon_and_name_title)
                .setPositiveButton(R.string.dialog_delete_positive_button, (dialog, which) ->
                        ShortcutsUtils.addPinnedShortcut(mEntity,
                                mBuilder.ivIcon.getDrawable(), mBuilder.etName.getText().toString()))
                .setNegativeButton(R.string.dialog_delete_negative_button, null)
                .create();
    }

    private void initView() {
        mBuilder.ivIcon.setImageDrawable(mIcon);
        mBuilder.etName.setText(mName);

        mBuilder.ivIcon.setOnClickListener(v -> {
            Intent intent = new Intent(Intent.ACTION_GET_CONTENT);
            intent.addCategory(Intent.CATEGORY_OPENABLE);
            intent.setType("image/*");
            startActivityForResult(intent, Const.REQUEST_CODE_IMAGE_CAPTURE);
        });
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        if (requestCode == Const.REQUEST_CODE_IMAGE_CAPTURE) {
            if (resultCode == RESULT_OK) {
                if (data != null) {
                    Uri iconUri = data.getData();
                    if (iconUri != null) {
                        Glide.with(this)
                                .load(iconUri)
                                .transition(DrawableTransitionOptions.withCrossFade())
                                .into(mBuilder.ivIcon);
                    }
                }
            }
            super.onActivityResult(requestCode, resultCode, data);
        }
    }
}

File: app/src/main/java/com/absinthe/anywhere_/ui/shortcuts/CreateShortcutDialogFragment.kt
package com.absinthe.anywhere_.ui.shortcuts

import android.app.Dialog
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.graphics.drawable.Drawable
import android.os.Bundle
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import com.absinthe.anywhere_.R
import com.absinthe.anywhere_.constants.GlobalValues
import com.absinthe.anywhere_.model.database.AnywhereEntity
import com.absinthe.anywhere_.utils.AppUtils
import com.absinthe.anywhere_.utils.ShortcutsUtils
import com.absinthe.anywhere_.utils.UxUtils
import com.absinthe.anywhere_.view.app.AnywhereDialogBuilder
import com.absinthe.anywhere_.view.app.AnywhereDialogFragment
import com.absinthe.anywhere_.viewbuilder.entity.CreateShortcutDialogBuilder
import com.absinthe.libraries.utils.extensions.dp
import com.blankj.utilcode.util.Utils
import com.bumptech.glide.Glide
import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions

class CreateShortcutDialogFragment : AnywhereDialogFragment() {
  private lateinit var mBuilder: CreateShortcutDialogBuilder
  private lateinit var imageResultLauncher: ActivityResultLauncher<String>
  private val mEntity: AnywhereEntity by lazy { arguments?.getParcelable(EXTRA_ENTITY) ?: AnywhereEntity() }
  private val mIcon: Drawable = UxUtils.getAppIcon(Utils.getApp(), mEntity, 45.dp)
  private val mName: String = mEntity.appName

  override fun onAttach(context: Context) {
    super.onAttach(context)
    imageResultLauncher = registerForActivityResult(ActivityResultContracts.GetContent()) {
      it?.let { uri ->
        activity?.let { activity ->
          Glide.with(activity.applicationContext)
            .load(uri)
            .transition(DrawableTransitionOptions.withCrossFade())
            .into(mBuilder.ivIcon)
        }
        AppUtils.takePersistableUriPermission(requireContext(), it, Intent())
      }
    }
  }

  override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
    mBuilder = CreateShortcutDialogBuilder(requireContext())
    val builder = AnywhereDialogBuilder(requireContext())
    initView()

    return builder.setView(mBuilder.root)
      .setTitle(R.string.dialog_set_icon_and_name_title)
      .setPositiveButton(R.string.dialog_delete_positive_button) { _: DialogInterface?, _: Int ->
        if (AppUtils.atLeastO() && !GlobalValues.deprecatedScCreatingMethod) {
          ShortcutsUtils.addPinnedShortcut(
            mEntity,
            mBuilder.ivIcon.drawable, mBuilder.etName.text.toString()
          )
        } else {
          ShortcutsUtils.addHomeShortcutPreO(
            mEntity,
            mBuilder.ivIcon.drawable, mBuilder.etName.text.toString()
          )
        }
      }
      .setNegativeButton(android.R.string.cancel, null)
      .create()
  }

  private fun initView() {
    mBuilder.apply {
      etName.setText(mName)
      ivIcon.apply {
        setImageDrawable(mIcon)
        setOnClickListener {
          imageResultLauncher.launch("image/*")
        }
      }
    }
  }

  companion object {

    const val EXTRA_ENTITY = "EXTRA_ENTITY"

    fun newInstance(entity: AnywhereEntity): CreateShortcutDialogFragment {
      return CreateShortcutDialogFragment().apply {
        arguments = Bundle().apply {
          putParcelable(EXTRA_ENTITY, entity)
        }
      }
    }
  }
}
